### PR TITLE
Rewrite the synonym endpoint

### LINKED ISSUE #389: [API] Synonym composed of many words
2 issues in [this route](https://docs.meilisearch.com/references/synonyms.html#delete-a-synonym), maybe linked:
- even if the synonym does not exist, it returns a `202 - Accepted` and an `updateId`, it sends it to the queue and marks it as `processed` -> ~~should return a `404` with a body error message.~~ OK
- if the synonym exists and is composed of many words, it sends it to the queue (expected behavior), marks it as `processed` but nothing happened.

### UPDATE:
I noticed that, in general, the routes do not work when the synonym is composed of many words.
These routes include: `get`, `update`, and `delete` for `/indexes/:uid/synonyms/:synonyms` and `:synonyms` is in two or more words.
- `get`: ~~crash, see this [issue](https://github.com/meilisearch/MeiliSearch/issues/386)~~ crash solved. But still does not work, see this [issue](https://github.com/meilisearch/MeiliSearch/issues/392)
- `update`: does not update and create a new synonym with the `%20` inside. (So, The update route doesn't only update, it can also create a new synonym, is this behavior expected?)
- `delete`: does not delete.

### LINKED ISSUE #392: Strange behavior on the get one synonym route
With this synonyms:
```json
{
    "smartphone": [
        "iphone",
        "samsung"
    ],
    "harry potter": [
        "hp"
    ],
    "hp": [
        "harry potter"
    ]
}
```

When I request:
- `GET /indexes/:uid/synonyms/hp`: OK
- `GET /indexes/:uid/synonyms/lol`: returns an empty array and `200`. Can we do the same behavior than the documents routes, that returns a `404` and a clear error message in the body?
- `GET /indexes/:uid/synonyms/harry%20potter`: returns an empty array and a `200`, but should return `[ "hp" ]`. Related to this [issue](https://github.com/meilisearch/MeiliSearch/issues/389).

### LINKED ISSUE #398: Rewrite the synonym endpoint
This issue is made to simplify the current endpoint of synonym. It's a temporary solution before we find a better way to do it. 

## Current Behavior

We have many routes. Here is an exhaustive list.

- List synonyms of one sequence `GET /indexes/:uid/synonyms/:synonym`
- List all sequences and synonyms `GET /indexes/:uid/synonyms`
- Create synonyms `POST /indexes/:uid/synonyms`
- Update a synonym `PUT /indexes/:uid/synonyms/:synonym`
- Delete a synonym `DELETE /indexes/:uid/synonyms/:synonym`
- Batch write synonyms `POST /indexes/:uid/synonyms/batch`
- Clear synonyms `DELETE /indexes/:uid/synonyms`

It was possible to create synonyms of two types differently. One-way synonyms and multi-way synonyms.
- One-way synonyms
```json
{
   "input":"magician",
   "synonyms":[
      "harry potter",
      "merlin"
   ]
}
```

- Multi-way synonyms
```json
{
   "synonyms":[
      "harry potter",
      "hp"
   ]
}
```

But the only way to get it back was in raw format. 

```json
{
    "smartphone": [
        "iphone",
        "samsung"
    ],
    "harry potter": [
        "hp"
    ],
    "hp": [
        "harry potter"
    ]
}
```

## Proposal

Limit the number of routes in the endpoint to two.

- Get synonyms `GET /indexes/:uid/synonyms`
- Update synonyms `POST /indexes/:uid/synonyms`

Only the raw format will be used in these two routes.
```json
{
    "smartphone": [
        "iphone",
        "samsung"
    ],
    "harry potter": [
        "hp"
    ],
    "hp": [
        "harry potter"
    ]
}
```
On synonym update the process will be: 
- Clear all current synonym
- Apply all synonyms given in body

This will also resolve issues #392 and #389

### PR DESCRIPTION
**Pull request by qdequele**
Related to #398 

closes #398, closes #392, closes #389 

Test:

Search "Green" > Only result with "Green"
Capture d’écran 2019-12-12 à 12 40 15

Add to synonyms "Green = Blue"
Capture d’écran 2019-12-12 à 12 40 30

Check synonyms list. "Green = Blue"
Capture d’écran 2019-12-12 à 12 42 45

Search "Green" > Result with "Green" and "Blue"
Capture d’écran 2019-12-12 à 12 40 41

Remove all synonyms
Capture d’écran 2019-12-12 à 12 40 53

Search "Green" > Only result with "Green"
Capture d’écran 2019-12-12 à 12 41 03

### CODE REVIEW

File: meilisearch-core/src/update/synonyms_update.rs [hunk 1]
```diff
@@ -56,21 +56,21 @@ impl SynonymsAddition {
...
     updates_store.put_update(writer, last_update_id, &update)?;
 
     Ok(last_update_id)
 }
 
-pub fn apply_synonyms_addition(
+pub fn apply_synonyms_update(
     writer: &mut heed::RwTxn<MainT>,
     main_store: store::Main,
     synonyms_store: store::Synonyms,
```
Comment (Kerollmops) [inline meilisearch-core/src/update/synonyms_update.rs]: You probably forget to `synonyms_store.clear()` before adding the new ones.

File: meilisearch-core/src/update/synonyms_update.rs [hunk 2]
```diff
@@ -92,28 +92,11 @@ pub fn apply_synonyms_addition(
         synonyms_store.put_synonyms(writer, word.as_bytes(), &alternatives)?;
```
Comment (Kerollmops) [inline meilisearch-core/src/update/synonyms_update.rs]: Here you put the alternatives to a word but you did not clear the previous ones.

File: meilisearch-core/src/update/synonyms_update.rs (hunk 1 above)
Comment (Kerollmops) [reply meilisearch-core/src/update/synonyms_update.rs]: The synonyms fst is just the list of all words that have alternatives, I am talking about the LMDB database (the kv store) that stores all the alternatives has the value of the key (the word that have alternatives).

File: meilisearch-core/src/update/synonyms_update.rs (hunk 2 above)
Comment (qdequele) [reply meilisearch-core/src/update/synonyms_update.rs]: Ok ! Got it, I will fix it !

**Comment by Kerollmops** [comment]
Nice, thank you for this PR, merging!

**Comment by bidoubiwa** [comment]
No documentation needed

**Pull request merged**